docs/source/triggers/index.rst

```diff
@@ -0,0 +1,271 @@
+   :end-before: # trigger-dependencies-end
+
+In addition to that you also need to import the ``Daml.Trigger``
+module.
+
+DAML triggers automatically track the active contract set and the
+commands in flight for you. In addition to that, they allow you to
+have user-defined state that is updated based on new transactions and
+command completions. For our copy trigger, the ACS is sufficient, so
+we will simply use ``()`` as the type of the user defined state.
+
+To create a trigger you need to define a value of type ``Trigger s`` where ``s`` is the type of your user-defined state:
+
+.. code-block:: daml
+
+    data Trigger s = Trigger
+      { initialize : ACS -> s
+      , updateState : ACS -> Message -> s -> s
+      , rule : Party -> ACS -> Map CommandId [Command] -> s -> TriggerA ()
+      }
+
+The ``initialize`` function is called on startup and allows you to
+initialize your user-defined state based on the active contract set.
+
+The ``updateState`` function is called on new transactions and command
+completions and can be used to update your user-defined state based on
+the ACS and the transaction or completion. Since our DAML trigger does
+not have any interesting user-defined state, we will not go into
+details here.
+
+Finally, the ``rule`` function is the core of a DAML trigger. It
+defines which commands need to be send to the ledger based on the
+party the trigger is executed at, the current state of the ACS, the
+commands in flight and the user defined state. The type ``TriggerA``
+allows you to emit commands that are then sent to the ledger. Like
+``Scenario`` or ``Update``, you can use ``do`` notation with
+``TriggerA``.
+
+Once you have defined the trigger, you have to pass it to ``runTrigger``. For our DAML trigger,
+this looks as follows:
+
+.. literalinclude:: ./template-root/src/CopyTrigger.daml
+   :language: daml
+   :start-after: -- TRIGGER_BEGIN
+   :end-before: -- TRIGGER_END
+
+Now we can move on to the most complex part of our DAML trigger, the implementation of ``copyRule``.
+First let’s take a look at the signature:
+
+.. literalinclude:: ./template-root/src/CopyTrigger.daml
+   :language: daml
+   :start-after: -- RULE_SIGNATURE_BEGIN
+   :end-before: -- RULE_SIGNATURE_END
+
+We will need the party and the ACS to get the ``Original`` contracts
+where we are the owner, the ``Subscriber`` contracts where we are in
+the ``subscribedTo`` field and the ``Copy`` contracts where we are the
+``owner`` of the corresponding ``Original``.
+
+The commands in flight will be useful to avoid sending the same
+command multiple times if ``copyRule`` is run multiple times before we
+get the corresponding transaction. Note that DAML triggers are
+expected to be designed such that they can cope with this, e.g., after
+a restart or a crash where the commands in flight do not contain
+commands in flight from before the restart, so this is an optimization
+rather than something required for them to function correctly.
```

Comment:
I find this sentence hard to parse. Maybe something like

Note that DAML triggers are expected to be robust against missing entries in the commands in flight. In particular they are not repopulated after a crash or restart. So, this is an optimization and should not be required for a trigger to function correctly.

Reply:
This changed after I merged @hurryabit’s suggestions. Does the new formulation help?

Reply:
Yes, I think it's easier to read in https://github.com/digital-asset/daml/pull/3183.